### PR TITLE
[SYCL] Meet queue capabilities on how large pattern fill it can process.

### PR DESCRIPTION
**Pull request by smaslov-intel**
Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -4773,10 +4777,23 @@ enqueueMemFillHelper(pi_command_type CommandType, pi_queue Queue, void *Ptr,
   // Get a new command list to be used on this call
   ze_command_list_handle_t ZeCommandList = nullptr;
   ze_fence_handle_t ZeFence = nullptr;
+
   // Performance analysis on a simple SYCL data "fill" test shows copy engine
   // is faster than compute engine for such operations.
+  bool PreferCopyEngine = Queue->Device->hasCopyEngine();
+  if (PreferCopyEngine) {
+    // Make sure that pattern size matches the capability of the queue.
+    if (PatternSize >
```
Comment (jandres742) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4786]: +1

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -4773,10 +4777,23 @@ enqueueMemFillHelper(pi_command_type CommandType, pi_queue Queue, void *Ptr,
   // Get a new command list to be used on this call
   ze_command_list_handle_t ZeCommandList = nullptr;
   ze_fence_handle_t ZeFence = nullptr;
+
   // Performance analysis on a simple SYCL data "fill" test shows copy engine
   // is faster than compute engine for such operations.
+  bool PreferCopyEngine = Queue->Device->hasCopyEngine();
+  if (PreferCopyEngine) {
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4784]: Two questions:
1. Why are we checking only when we want to use Copy Engine? Should we not be checking this irrespective of whether we want to use copy or compute engine?
2. Also, Are we checking if the pattern size is a power of two anywhere?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -4773,10 +4777,23 @@ enqueueMemFillHelper(pi_command_type CommandType, pi_queue Queue, void *Ptr,
...
+    // Make sure that pattern size matches the capability of the queue.
+    if (PatternSize >
+        Queue->Device->ZeCopyQueueGroupProperties.maxMemoryFillPatternSize) {
+      PreferCopyEngine = false;
+      PI_ASSERT(PatternSize <= Queue->Device->ZeComputeQueueGroupProperties
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4789]: I think this should be 'less than' based on doc.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.cpp:4786]: This should be >=

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:4784]: 1. Here I am mostly checking if to use copy engine or not. I can rework the code that the assertion that the chosen queue can support the size is performed always.
2. Don't know, will add the check.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:4789]: Level-Zero header file says:
``` C++
size_t maxMemoryFillPatternSize;                ///< [out] maximum `pattern_size` supported by command queue group.
```

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: we actually had the pow-2 check already later in the code, moved it

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (asudarsa) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Hmm..the SPEC doc says
The pattern size must be a power-of-two and less than ze_command_queue_group_properties_t.maxMemoryFillPatternSize.
I am looking at the SPEC doc from here:
https://spec.oneapi.com/level-zero/latest/core/api.html#ze__api_8h_1a689ec654289f637226b9d6e84a42473b
Thanks

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Right, there is a disconnect between the header and the spec.
In my case, implementation reports 16 for maximum on copy-queue, and given that it must also be a power of 2, it does not make sense to be "less than 16".

I am clarifying with the spec owners.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: We want to "max" be inclusive value, and will adjust the spec. So code here is good as is.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -604,6 +604,12 @@ pi_result _pi_context::finalize() {
     if (ZeCommandList)
       ZE_CALL(zeCommandListDestroy, (ZeCommandList));
   }
+
```
Comment (asudarsa) [inline sycl/plugins/level_zero/pi_level_zero.cpp:607]: Just curious, why is this required?

**Review by asudarsa — DISMISSED** [review]
Looks good to me. Thanks

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 4 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:607]: Otherwise the global count always grows and soon reaches the limit of max created command-lists

**Review by asudarsa — APPROVED** [review]
